magma/lib/magma/loader.rb

```diff
@@ -134,6 +136,91 @@ def explicit_revision_exists?(revisions:, parent_model:, parent_record_name:, ch
         child_record_name)
     end
 
+    def find_parent_models(identifier)
+      # This method attempts to find all parent models and (their identifiers) for a given identifier.
+      #
+      # It returns a hash of hashes, where each hash is indexed by model name, and each sub hash, contains the model,
+      # identifier_name, identifier and the parent_model. If a identifier does not exist in the hierarchy chain, an
+      # empty list is returned.
+      #
+      # This method succeeds at finding parent identifiers if :
+      # 1. A proper grammar is created with gnomon and the grammar rules correspond to a model hierarchy.
+      # 2. All models have been created in the db.
+      #
+      # The identifiers DO NOT need to exist in the identifier table.
+
+      return {} if @grammar.nil?
+      grammar_decomposed = @grammar.decompose(identifier)
+      return {} if grammar_decomposed.nil?
+
+      # First we gather all available parent models and identifiers associated with the identifier.
+      # If someone has created a well formatted grammar, with proper identifiers,
+      # there MAY be a model hierarchy chain we can infer
+
+      project = Magma.instance.get_project(@project_name)
+      available_models = project.models.keys & grammar_decomposed[:rules].keys.map(&:to_sym)
+
+      chain = {}
+      available_models.each do |model_name|
+        model = project.models[model_name]
+        chain[model_name] = {
+            model: model,
+            identifier_name: model.identity.name,
+            identifier: grammar_decomposed[:rules][model_name.to_s][:name],
+            parent_model_name: model.parent_model_name
+          }
+      end
+
+      # Then we check to make sure a valid hierarchy chain exists
+      # We check that each parent_model is in the list of models available
+      arr = chain.to_a
+      valid = arr.all? do |el|
+        model_name = el[0]
+        model_name == :project || arr.any? do |other|
+          candidate_model_name = other[0]
+          el[1][:parent_model_name] == candidate_model_name
+        end
+      end
```

Comment:
Because Hash < Enumerable, could do like:
```
valid = chain.all? do |model_name, model|
  model_name == :project || chain.has_key?(model[:parent_model_name])
end
```

Reply:
nice, much cleaner